connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java

```diff
@@ -712,8 +733,16 @@ KafkaBasedLog<String, byte[]> setupAndCreateKafkaBasedLog(String topic, final Wo
     }
 
     private void sendPrivileged(String key, byte[] value) {
+        sendPrivileged(key, value, null);
+    }
+
+    private void sendPrivileged(String key, byte[] value, Callback<Void> callback) {
         if (!usesFencableWriter) {
-            configLog.send(key, value);
```

Comment:
The only other call-site for the send(String, byte[]) is now putTargetState, does a similar fix apply there?
Perhaps this blind send call is not safe and should be eliminated. Perhaps we can also enforce non-null callbacks. WDYT?

Reply:
Thanks, that's a good observation. Currently, the pause / resume APIs (which call `putTargetState`) don't use callback mechanisms whatsoever and they also don't have any documented response bodies (they send back a `202` response with an empty body right now) and will hence need a bit more refactoring. I feel like that should be done in a follow up / separate PR, WDYT?

Reply:
> Edit 2: Although now with these changes, there's an inconsistency between different APIs since deleteConnectorConfig, putTaskConfigs and a couple of other methods are still using null callbacks.

Rather than rewriting all of the `ConfigBackingStore` methods to accept callbacks (making them look async), maybe we can terminate the producer callbacks in the `KafkaConfigBackingStore` and keep the `ConfiBackingStore` methods synchronous. I think this has other advantages, especially for the `putConnectorConfig`, where we can enforce that the producer send completes before the `readToEnd()`.

It appears that the original signatures for the `ConfigBackingStore` provided a synchronous API. The javadocs don't mention what exceptions are thrown, but there are some methods which already throw exceptions, such as those coming from the readToEnd() calls. We could certainly throw exceptions rather than propagate them via callbacks.

In fact, the only `*BackingStore` method which was already async was `OffsetBackingStore::set(Map, Callback)`, which appears was only written that way to log errors (no change in control flow, retries, etc). That method is now involved in some callback hell in ConnectorOffsetBackingStore used for EOS source that could also be rewritten to be synchronous. (I don't think this is in-scope for this PR and is just an interesting lateral refactor)

Reply:
> if we want to surface producer errors to users ... which would involve changing the method signatures to accept callbacks

This is not true; you can surface errors synchronously by throwing exceptions inside the `KafkaConfigBackingStore`  and catching them in the caller. This is what I meant by terminating the producer callbacks: transform the control flow from asynchronous via a callback to synchronous results.

Reply:
Oh okay, I see what you're saying now. Sorry, I think I misunderstood you earlier. I agree with you, none of the `KafkaConfigBackingStore` use cases seem to necessarily require the performance benefit of using async send with callbacks. Although to make it synchronous, I think it might be better to avoid using the producer callback altogether and instead call `get()` on the returned future (which throws any exceptions that were encountered while sending the record), WDYT?

Reply:
I've gone ahead and made the changes to convert the `KafkaConfigBackingStore` APIs to be synchronous even when EOS is disabled (thus making the behavior more in-line with the EOS case). This simplifies things significantly while also providing consistency w.r.t error handling across all the `KafkaConfigBackingStore` APIs without making invasive changes.

Reply:
Do we still see the cause of the failure to send in the REST response with the current changes? It looks like we're only going to include the generic "Failed to... to/from Kafka" message, or the privileged write failure message.

I think it'd be nice to include more detail on the cause of the failure, but if that's too invasive or difficult to get quite right, it can be left as a follow-up item.

Reply:
(Resolving this convo as it's taking up way too much real estate in the GitHub UI)